starky/src/permutation_stark.rs

```diff
@@ -0,0 +1,238 @@
+//! An example of generating and verifying a STARK to highlight the use of the
+//! permutation argument with logUp.
+
+#[cfg(not(feature = "std"))]
+use alloc::{vec, vec::Vec};
+use core::marker::PhantomData;
+
+use plonky2::field::extension::{Extendable, FieldExtension};
+use plonky2::field::packed::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::evaluation_frame::StarkFrame;
+use crate::lookup::{Column, Lookup};
+use crate::stark::Stark;
+use crate::util::trace_rows_to_poly_values;
+
+/// Computes a sequence with state `[i, j]` using the state transition
+/// i' <- i+1, j' <- j+1`.
+/// Note: The `0, 1` columns are the columns used to test the permutation argument.
+#[derive(Copy, Clone)]
+struct PermutationStark<F: RichField + Extendable<D>, const D: usize> {
+    num_rows: usize,
+    _phantom: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> PermutationStark<F, D> {
+    const fn new(num_rows: usize) -> Self {
+        Self {
+            num_rows,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Generate the trace using `x0, x1, 0, 1, 1` as initial state values.
+    fn generate_trace(&self, x0: F) -> Vec<PolynomialValues<F>> {
+        let mut trace_rows = (0..self.num_rows)
+            .scan([x0, x0 + F::ONE, F::ONE], |acc, _| {
+                let tmp = *acc;
+                acc[0] = tmp[0] + F::ONE;
+                acc[1] = tmp[1] + F::ONE;
+                // acc[2] (i.e. frequency column) remains unchanged, as we're permuting a strictly monotonous sequence.
+                Some(tmp)
+            })
+            .collect::<Vec<_>>();
+        trace_rows[self.num_rows - 1][1] = F::ZERO; // So that column 0 and 1 are permutation of one another.
```

Comment:
This passes fine because the test sets `x0` to `0`, but it should be 
```suggestion
        trace_rows[self.num_rows - 1][1] = x0; // So that column 0 and 1 are permutation of one another.
```